File: r/10.R
#!/usr/bin/env Rscript

library(magrittr)

look.say <- function(input) {
  r <- input %>% strsplit('') %>% unlist %>% rle
  rbind(as.character(r$lengths), r$values) %>% paste(collapse = '')
}

side.a <- function(input) {
  x <- input
  for(i in 1:40) {
    x <- look.say(x)
  }
  nchar(x)
}

side.b <- function(input) {
  x <- input
  for(i in 1:50) {
    x <- look.say(x)
  }
  nchar(x)
}

main <- function() {
    input <- '3113322113'
    cat(side.a(input), '\n')
    cat(side.b(input), '\n')
}

main()

File: js/10.js
#!/usr/bin/env node
'use strict'

const fs = require('fs')

const rle = (str) => str
  .replace(/(.)\1*/g, (seq, pl) => seq.length.toString() + pl)

const sideA = (str) => {
  let x = str
  for (let i = 0; i < 40; ++i) {
    x = rle(x)
  }
  return x.length
}

const sideB = (str) => {
  let x = str
  for (let i = 0; i < 50; ++i) {
    x = rle(x)
  }
  return x.length
}

const main = () => {
  const input = fs.readFileSync('../input/10.txt', 'utf8').trim()
  console.log(sideA(input))
  console.log(sideB(input))
}

main()
